### PR TITLE
Refactor Doughnut for 2.0 Changes

### LINKED ISSUE #46: Add a new `Delegated` origin variant (currently, signed, none, root). Pass the doughnut through to module methods using this delegated origin type. This will allow us to remove the doughnut context storage hack (see: https://github.com/cennznet/plug-blockchain/blob/8d56aa31981d6944eadccbb5f7f430333f7dedaa/srml/support/src/dispatch.rs#L1061, https://github.com/cennznet/plug-blockchain/blob/8d56aa31981d6944eadccbb5f7f430333f7dedaa/srml/executive/src/doughnut/mod.rs#L173-L179)
##21 points
Relevant conversion trait bounds will require update to use `Origin: From<Option<AccountId>, Option<Doughnut>>`

### PR DESCRIPTION
**Pull request by jordy25519**
## Change Summary
- Introduce `RawOrigin::Delegated(AccountId, Doughnut)` to allow passing the doughnut proof to runtime methods and ensuring delegation is known at all times by the runtime.
- Use the `SignedExtension` feature of transactions to handle doughnut logic, removed form executive

Closes #46 and 90% of groundwork for https://github.com/cennznet/cennznet/issues/54

## TODO
- [x] Fix srml/support tests
- [x] Reinstate executive tests?
- [x] Integration tests for delegated transactions
- [ ] Fix `prml-doughnut` tests with `cargo test --all`

## ChangeLog
Files: `core/sr-primitives/src/traits.rs`

- Blanket `impl SignedExtension for Option<T>`
- `MaybeDoughnut` trait for extracting doughnut from SignedExtra tuple
- impl MaybeDoughnut for SignedExtension macro tuple of all lengths
- Changed `trait SignedExtension::pre_dispatch` to take self by reference (`&self`), not move (`self`)

Files: `srml/system/src/lib.rs` and `srml/support/src/origin.rs`
- Add `DelegatedOrigin` variant (AccountId, Doughnut)
- Add `MaybeDoughnutRef` trait for extracting doughnut from origin without move in runtime methods
- Improve doughnut type bounds `type Doughnut: Parameter + Member + DoughnutApi;`

Files: `srml/executive`
- Remove DoughnutExecutive, it has been superseded by the hooks provided by `SignedExtension`trait.  

Files: `node/runtime/src/lib.rs`
- DelegatedOrigin type, it's now necessary as the context storage hack is outside of the executive/runtime and won't work anymore
- Removed `node/runtime/src/doughnut.rs`, moved into `prml/doughnut`

Files: `prml/doughnut`
Adds:
- Add PlugDoughnut as wrapper struct for doughnut to use new tx `SignedExtension`
- Add PlugDoughnutDispatcher and impl for DispatchVerifier
- impl DoughnutVerify for PlugDoughnut
- impl SignedExtension for PlugDoughnut

Files: `node/primitives/plug_extrinsic.rs`
- Removed, now using `SignedExtension` functionality

### CODE REVIEW

File: srml/support/test/tests/instance.rs [hunk 1]
```diff
@@ -1,516 +1,516 @@
...
-fn instance_prefix_is_prefix_of_entries() {
-	use module2::Instance;
-
-	let prefix = module2::Instance2::PREFIX;
-	assert!(module2::Instance2::PREFIX_FOR_Value.starts_with(prefix));
-	assert!(module2::Instance2::PREFIX_FOR_Map.starts_with(prefix));
-	assert!(module2::Instance2::PREFIX_FOR_LinkedMap.starts_with(prefix));
-	assert!(module2::Instance2::PREFIX_FOR_DoubleMap.starts_with(prefix));
-}
+// // Copyright 2019 Parity Technologies (UK) Ltd.
```
Comment (jordy25519) [inline srml/support/test/tests/instance.rs]: This test works correctly using `cargo test -p srml-support-test` but fails with `cargo test --all`, has been tricky to debug. Commented out for the sake of progress

File: core/sr-api-macros/Cargo.toml [hunk 2]
```diff
@@ -25,6 +25,7 @@ criterion = "0.2"
 consensus_common = { package = "substrate-consensus-common", path = "../consensus/common" }
 codec = { package = "parity-scale-codec", version = "1.0.0" }
 trybuild = "1.0"
+protobuf = "~2.8.0"
```
Comment (jordy25519) [inline core/sr-api-macros/Cargo.toml:30]: Locked as protobuf update broke semver and breaks libp2p-core build. See https://github.com/paritytech/substrate/pull/3723/files

File: prml/doughnut/src/impls.rs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+			signature_version: 0,
+			domains: vec![("test".to_string(), vec![0u8])],
+		};
+		doughnut.signature = issuer.sign(&doughnut.payload()).into();
+		let plug_doughnut = PlugDoughnut::<_, Runtime>::new(doughnut);
+		assert!(plug_doughnut.verify());
+	}
+
+	#[test]
+	fn plug_doughnut_does_not_verify_sr25519_signature() {
```
Comment (elim0322) [inline prml/doughnut/src/impls.rs]: this test is failing on my machine

File: prml/doughnut/src/impls.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+			signature_version: 0,
+			domains: vec![("test".to_string(), vec![0u8])],
+		};
+		doughnut.signature = holder.sign(&doughnut.payload()).into();
+		let plug_doughnut = PlugDoughnut::<_, Runtime>::new(doughnut);
+		assert_eq!(plug_doughnut.verify(), false);
+	}
+
+	#[test]
+	fn plug_doughnut_verifies_ed25519_signature() {
```
Comment (cowboy-bebug) [inline prml/doughnut/src/impls.rs:234]: this test fails on my machine

File: prml/doughnut/src/impls.rs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+			<PlugDoughnut<_, _> as DoughnutApi>::validate(&plug_doughnut, holder.public(), 3001).is_err()
+		);
+		// signer is not holder
+		assert!(
+			<PlugDoughnut<_, _> as DoughnutApi>::validate(&plug_doughnut, signer.public(), 100).is_err()
+		);
+	}
+
+	#[test]
+	fn plug_doughnut_verifies_sr25519_signature() {
```
Comment (cowboy-bebug) [inline prml/doughnut/src/impls.rs:196]: this test fails on my machine

File: prml/doughnut/src/impls.rs (hunk 4 above)
Comment (jordy25519) [reply prml/doughnut/src/impls.rs:234]: I'm trying to fix these at the moment. It's a weird build error, it works if you run it like `cargo test -p prml-doughnut`

**Comment by shaunxw** [comment]
Has CI been turned off?

**Comment by cowboy-bebug** [comment]
> Has CI been turned off?

I think it's running. Looks like it's been running for > 1 hr 😄

**Comment by shaunxw** [comment]
> > Has CI been turned off?
> 
> I think it's running. Looks like it's been running for > 1 hr 😄

Wired, didn't see any CI building status on this page - -

**Review by shaunxw — APPROVED** [review]
LGTM

**Comment by cowboy-bebug** [comment]
@holygits reasons why the tests are failing could be due to version mismatch between doughnut and the substrate. For example `schnorrkel v0.1.x` for doughnut vs `0.8.4` on substrate. I did notice our tests failing on doughnut-rs when I bumped up this crate

https://github.com/cennznet/doughnut-rs/issues/11#issuecomment-540338294

**Comment by jordy25519** [comment]
Fixed the tests in this branch + substrate updates from #49 : https://github.com/cennznet/plug-blockchain/tree/develop

**Review by cowboy-bebug — APPROVED** [review]

**Comment by jordy25519** [comment]
Superseded by https://github.com/cennznet/plug-blockchain/tree/develop

**Pull request closed without merging**